pkg/framework/runner.go

```diff
@@ -294,17 +295,33 @@ func (r *runner) evaluatePredicates(world map[string]interface{}) {
 		worldCopy := runtime.DeepCopyJSON(world)
 		iter := code.Run(worldCopy, worldCopy)
 		hadElem := false
+		var errMsgs []string
 		for result, ok := iter.Next(); ok; result, ok = iter.Next() {
 			hadElem = true
 			err, _ := result.(error)
 			if errors.Is(err, logic.ErrAssumptionViolation) {
 				continue
 			}
-			if !r.Assert().NoErrorf(err, "failed to evaluate pred %q", pred) {
-				continue
+			if err != nil {
+				errMsgs = append(errMsgs, err.Error())
+			} else if !logic.Truthy(result) {
+				errMsgs = append(errMsgs, fmt.Sprintf("evaluated to falsy result %v", result))
 			}
-			r.Assert().True(logic.Truthy(result), "predicate %q evaluated to falsy result %v", pred, result)
 		}
-		r.Assert().Truef(hadElem, "predicate %q evaluated to empty sequence", pred)
+		if !hadElem {
+			errMsgs = append(errMsgs, "evaluated to empty sequence")
+		}
+		if len(errMsgs) == 0 {
+			continue
+		}
+		fullMsg := "predicate"
+		if sctx := pred.SourceCtx; !sctx.IsZero() {
+			fullMsg += fmt.Sprintf(" at %s", sctx)
+		}
+		fullMsg += fmt.Sprintf(":\n  %s\nerrors:\n", pred.Source)
+		for _, errMsg := range errMsgs {
+			fullMsg += fmt.Sprintf("  %s\n", errMsg)
+		}
+		r.Assert().Failf(fullMsg, "error(s) evaluating predicate %q", pred)
```

Comment:
Can you add a small (integration) test case to see that the message is rendered correctly?

Reply:
I have to look into that. Is it possible to assert that something fails an assertion? I don't really want to mock the whole `Assert` interface..

Reply:
Yeah, I think this is not feasible to do in a shorter timeframe. The problem is that the whole `runner` type and friends heavily rely on `testing.T`, which is not mockable. I also cannot rewrite it to depend on `require.TestingT` from `testify`, because I do net `t.Run(...)`. And because Go doesn't allow contravariant parameter types in interface implementations, it is principally _impossible_ to define an `interface` that is (a) satisfied by `*testing.T` and (b) independent of the non-mockable `*testing.T`.

A proper integration test would use `bats` or so, which I consider overkill for this.